### ISSUE TITLE
ACTNARS fails to resolve closed generics

### ISSUE THREAD

**Issue by droyad** [autofac/Autofac#855]
We are using `AnyConcreteTypeNotAlreadyRegisteredSource`, and it's behaviour has changed in 4.2.0 due to [this](https://github.com/autofac/Autofac/commit/1aca241fc0e7d5535fdff23c833dcbfdaa0f02b7) change stemming from #495.

For example we want to resolve `Foo<Bar>`  where we have an implementation `Foo<T>`. The `ACTNARS` delegate is `t => t.Name.StartsWith("Foo")`. `Foo<T>` does not take `Bar` in the constructor.

`ACTNARS` currently (`4.6.0`) fails to resolve this type as it checks that `Bar` is registered ([here](https://github.com/autofac/Autofac/blob/v4.6.0/src/Autofac/Features/ResolveAnything/AnyConcreteTypeNotAlreadyRegisteredSource.cs#L98)

```
     [Fact]
        public void ConstructableOpenGenericsCanBeResolved()
        {
            var cb = new ContainerBuilder();
            cb.RegisterSource(new AnyConcreteTypeNotAlreadyRegisteredSource(t => t.Name.StartsWith("Progress")));
            var container = cb.Build();
            Assert.True(container.IsRegistered<Progress<Exception>>());
            Assert.NotNull(container.Resolve<Progress<Exception>>());
        }
```

**Comment by tillig** [comment]
Good find. We'll have to think about how best to solve this. `AnyConcreteTypeNotRegisteredSource` and `ContravariantRegistrationSource` are two great, flexible, pains in my behind. In #495 we found ACTNARS interfering with other registration sources but we really can't put special cases around `Meta<T>` or other source "known types." Maybe I fixed that wrong, or maybe there's more to check? Hmmm. Will have to dig in.
